### ISSUE TITLE
Absent __STDC_NO_VLA__ macro

### ISSUE THREAD

**Issue by ComdivByZero** [AbsInt/CompCert#221]
ISO C11 standard turn VLA into optional feature and introduce macro __STDC_NO_VLA__=1 for implementations, which are not supports Variable Length Arrays concept. 
Yes, CompCert support only subset of C99, but such feature very easy to implement and can be helpful.

Also, VLA is not only dangerous feature, when used to allocate local array, but very useful when used to inform about length of array -formal-parameter.

```
void func (int len, int array[len]) {
     assert ( len == sizeof(array)/sizeof(array[0]) );
}
```

What about partial implementation of VLA for formal parameters?

**Comment by xavierleroy** [comment]
There are two suggestions in this report:

1. Describe the CompCert subset of C by reference to ISO C 2011 instead of ISO C 1999, in which case we can just define STDC_NO_VLA=1 to explain that VLA are not supported.  I like this idea.  However we need to check with the industrial users of CompCert: they tend to use C99, or even C90, as the reference standard for qualification arguments.  So, I'm marking this as "enhancement".

2. Support VLAs in types of function parameters but not in types of local variables.  That's an original idea.  I'm afraid it's more confusing than useful for the programmers.  Is it the case that this "half-support" is allowed by the C2011 standard?  I don't think so.

**Comment by bschommer** [comment]
CompCert now defines some of the conditional feature macros of C11, including `__STDC_NO_VLA__`.

**Comment by mwallnoefer** [comment]
As a follow-up, C23 seems to have changed the semantics of the `__STD_NO_VLA__` macro. _Cppreference_ describes the situation as follows:

From (since C11)(until C23):
> If the compiler defines the macro constant __STDC_NO_VLA__ to integer constant 1, then VLA and VM types are not supported.

into (since C23):
> If the compiler defines the macro constant __STDC_NO_VLA__ to integer constant 1, then VLA objects with automatic storage duration are not supported.
> 
> The support for VM types (variable modified types = VLAs for formal parameters) and VLAs with allocated storage durations is mandated.

Source: https://en.cppreference.com/w/c/language/array#Explanation.

So maybe have another consideration for implementing this feature in an upcoming release.

**Comment by xavierleroy** [comment]
Yes, you're correct about VLAs making a come back in C23 in a limited form (stack allocation may not be supported), which makes them easier to support.  I haven't looked at supporting (as much as possible of) C23 in CompCert yet.  Right now we're still concentrating on C11 and C17.

**Comment by mwallnoefer** [comment]
Some very basic patch which enables the (unchecked) VLA support in formal parameters:

I have just changed the following two lines:
https://github.com/AbsInt/CompCert/blob/c7b27e30a9057c14d5a02a9f9aa3d8f230715eb0/cparser/Elab.ml#L913-L914
into
```
warning Flexible_array_extensions loc "size of array is not a compile-time constant";
None in (* produces better error messages later *)
```

Now the compilation happens as if _no_ length had been specified (`[]`). VLAs with stack allocation remain prohibited by means of the following additional error message:
> prog4.c:3: warning: size of array is not a compile-time constant [-Wflexible-array-extensions]
> prog4.c:3: error: variable 'a' has incomplete type 'int[]'

In other words the patch should not enable ccomp to generate unsafe code.

**Comment by xavierleroy** [comment]
Your patch should work for one-dimensional arrays
```
double f(int n, double a[n]) { return a[1]; }
```
but it's not going to handle multi-dimensional arrays, which are one of the interesting use cases for VLAs:
```
double f(int n, int m, double a[n][m]) { return a[1][2]; }
```
This should be rejected even after your patch, just like the declaration `double a[][]` is rejected.  It would be unsafe, however, if it was treated like `double * a[]`, i.e. an array of pointers to arrays...

While C23 VLAs remove the need for variable-length allocations in the stack, they still require array types to contain arbitrary expressions, not just constants, and that's what "goes against the grain" of the current CompCert code base.

**Comment by mwallnoefer** [comment]
Indeed, I had not tested any multi-dimensional arrays. My patch was just a small hack to overcome the _VLA error_ and I thought it could be safe. But your comment made me realise that this is not always the case, so yes, better stick to the existing behaviour.
